internal/services/v1/lookupwatch_test.go

```diff
@@ -0,0 +1,207 @@
+package v1_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	v1 "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"github.com/authzed/spicedb/internal/datastore/memdb"
+	"github.com/authzed/spicedb/internal/testfixtures"
+	"github.com/authzed/spicedb/internal/testserver"
+	lookupwatchv1 "github.com/authzed/spicedb/pkg/proto/lookupwatch/v1"
+	"github.com/authzed/spicedb/pkg/zedtoken"
+)
+
+func permissionUpdate(
+	resourceType string,
+	resourceId string,
+	subjectType string,
+	subjectId string,
+	permission v1.CheckPermissionResponse_Permissionship,
+) *lookupwatchv1.PermissionUpdate {
+	return &lookupwatchv1.PermissionUpdate{
+		Resource: &v1.ObjectReference{
+			ObjectType: resourceType,
+			ObjectId:   resourceId,
+		},
+		Subject: &v1.SubjectReference{
+			Object: &v1.ObjectReference{
+				ObjectType: subjectType,
+				ObjectId:   subjectId,
+			},
+		},
+		UpdatedPermission: permission,
+	}
+}
+
+func TestLookupWatch(t *testing.T) {
+	testCases := []struct {
+		name               string
+		resourceObjectType string
+		permission         string
+		startCursor        *v1.ZedToken
+		mutations          []*v1.RelationshipUpdate
+		expectedCode       codes.Code
+		expectedUpdates    []*lookupwatchv1.PermissionUpdate
+	}{
+		{
+			name:               "lookupWatch basic test relation CREATE",
+			expectedCode:       codes.OK,
+			resourceObjectType: "document",
+			permission:         "view",
+			mutations: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1", ""),
+			},
+			expectedUpdates: []*lookupwatchv1.PermissionUpdate{
+				permissionUpdate("document", "document1", "user", "user1", v1.CheckPermissionResponse_PERMISSIONSHIP_HAS_PERMISSION),
+			},
+		},
+		{
+			name:               "lookupWatch basic test relation change on an intermediate relation",
+			expectedCode:       codes.OK,
+			resourceObjectType: "document",
+			permission:         "view",
+			mutations: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_DELETE, "folder", "company", "viewer", "folder", "auditors", "viewer"),
+			},
+			expectedUpdates: []*lookupwatchv1.PermissionUpdate{
+				permissionUpdate("document", "companyplan", "user", "auditor", v1.CheckPermissionResponse_PERMISSIONSHIP_NO_PERMISSION),
+				permissionUpdate("document", "masterplan", "user", "auditor", v1.CheckPermissionResponse_PERMISSIONSHIP_NO_PERMISSION),
+			},
+		},
+		{
+			name:               "lookupWatch basic test with no change impact on the watched resourceType",
```

Comment:
We have an issue on this test (RR doesn't return any data)
But once we fix the next issue `lookupWatch adding an intermediate relation`, it will also fix this test

BTW the test doesn't report an error because we expect no updated permissions (hence RR returning no data has no impact here)